pkg/cloudprovider/openstack.go

```diff
@@ -237,47 +222,127 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+	}
+	// List all ports that are attached to this server.
+	serverID, err := getNovaServerIDFromProviderID(node.Spec.ProviderID)
+	if err != nil {
+		return err
+	}
+
+	matchingSubnet, matchingPort, err := o.findAssignSubnetAndPort(ip, node)
+	if err != nil {
+		return err
+	}
+
+	if matchingSubnet != nil {
+		// 2) Reserve the IP address on the subnet by creating a new unattached neutron port.
+		unboundPort, err := o.reserveNeutronIPAddress(*matchingSubnet, ip, serverID)
+		if err != nil {
+			return err
+		}
+		// 3) Then, add the IP address to the port's allowed_address_pairs.
+		//    TODO: use a more elegant retry mechanism.
+		if err = o.allowIPAddressOnNeutronPort(matchingPort.ID, ip); err != nil && !errors.Is(err, AlreadyExistingIPError) {
+			// Try to clean up the allocated port if adding the IP to allowed_address_pairs failed.
+			// Try this 10 times, but if this operation fails more than that, then user intervention is needed or
+			// the upper layer must call ReleasePrivateIP (because if the neutron port exists and holds
+			// a reservation, then the assign step will not continue after step 2).
+			var errRelease error
+			var releaseStatus string
+			for i := 0; i < 10; i++ {
+				errRelease = o.releaseNeutronIPAddress(*unboundPort, serverID)
+				// If the release operation was successful, then we are done.
+				if errRelease == nil {
+					releaseStatus = "Released neutron port reservation."
+					break
 				}
-				return fmt.Errorf("could not allow IP address %s on port %s, err: %q. %s", ip.String(), serverPort.ID, err, releaseStatus)
+				// Otherwise store the error message and retry.
+				releaseStatus = fmt.Sprintf("Could not release neutron port reservation after %d tries, err: %q", i+1, errRelease)
 			}
-			// 4) Return nil to indicate success if steps 2 and 3 passed.
-			return nil
+			return fmt.Errorf("could not allow IP address %s on port %s, err: %q. %s", ip.String(), matchingPort.ID, err, releaseStatus)
 		}
+		// 4) Return nil to indicate success if steps 2 and 3 passed.
+		return nil
 	}
 
 	// 5) The IP address does not fit in any of the attached networks' subnets.
 	return fmt.Errorf("could not assign IP address %s to node %s", ip, node.Name)
 }
 
+func (o *OpenStack) MovePrivateIP(ip net.IP, nodeToAdd, nodeToDel *corev1.Node) error {
+	if nodeToAdd == nil || nodeToDel == nil {
+		return fmt.Errorf("invalid nil pointer provided for node when trying to move IP %s", ip.String())
+	}
+
+	// List all ports that are attached to this server.
+	serverID, err := getNovaServerIDFromProviderID(nodeToDel.Spec.ProviderID)
+	if err != nil {
+		return err
+	}
+	serverPorts, err := o.listNovaServerPorts(serverID)
+	if err != nil {
+		return err
+	}
+
+	// Loop over all ports that are attached to this nova instance.
+	for _, serverPort := range serverPorts {
+		if isIPAddressAllowedOnNeutronPort(serverPort, ip) {
+			if err = o.unallowIPAddressOnNeutronPort(serverPort.ID, ip); err != nil {
```

Comment:
Does the comment in 332 refer to this line here? If so, I'd say we need to return the error, because we check if the IP is in the allowed_address_pairs, and only if that's the case, we try to remove it (and retry). So if this still fails, then something went wrong during the detach / update of allowed_address_pairs. Returning the error will make sure that we do not continue without removing the IP from the allowed_address_pair on the source port. This is also important because without this succeeding, findAssignSubnetAndPort might return AlreadyExistingIPError

Reply:
We do remove the error. The comment is about - what if for each of the ports `isIPAddressAllowedOnNeutronPort()` was false?